src/script/interpreter.cpp

```diff
@@ -1050,15 +1051,135 @@ class CTransactionSignatureSerializer {
     }
 };
 
+const unsigned char ZCASH_INNER_HASH_PERSONALIZATION[crypto_generichash_blake2b_PERSONALBYTES] =
+    {'Z','c','a','s','h','_','I','n','n','e','r','_','H','a','s','h'};
+
+uint256 GetPrevoutHash(const CTransaction& txTo) {
+    CBLAKE2bWriter ss(SER_GETHASH, 0, ZCASH_INNER_HASH_PERSONALIZATION);
+    for (unsigned int n = 0; n < txTo.vin.size(); n++) {
+        ss << txTo.vin[n].prevout;
+    }
+    return ss.GetHash();
+}
+
+uint256 GetSequenceHash(const CTransaction& txTo) {
+    CBLAKE2bWriter ss(SER_GETHASH, 0, ZCASH_INNER_HASH_PERSONALIZATION);
+    for (unsigned int n = 0; n < txTo.vin.size(); n++) {
+        ss << txTo.vin[n].nSequence;
+    }
+    return ss.GetHash();
+}
+
+uint256 GetOutputsHash(const CTransaction& txTo) {
+    CBLAKE2bWriter ss(SER_GETHASH, 0, ZCASH_INNER_HASH_PERSONALIZATION);
+    for (unsigned int n = 0; n < txTo.vout.size(); n++) {
+        ss << txTo.vout[n];
+    }
+    return ss.GetHash();
+}
+
+uint256 GetJoinSplitsHash(const CTransaction& txTo) {
+    CBLAKE2bWriter ss(SER_GETHASH, 0, ZCASH_INNER_HASH_PERSONALIZATION);
+    for (unsigned int n = 0; n < txTo.vjoinsplit.size(); n++) {
+        ss << txTo.vjoinsplit[n];
+    }
+    ss << txTo.joinSplitPubKey;
+    return ss.GetHash();
+}
+
 } // anon namespace
 
-uint256 SignatureHash(const CScript& scriptCode, const CTransaction& txTo, unsigned int nIn, int nHashType)
+PrecomputedTransactionData::PrecomputedTransactionData(const CTransaction& txTo)
+{
+    hashPrevouts = GetPrevoutHash(txTo);
+    hashSequence = GetSequenceHash(txTo);
+    hashOutputs = GetOutputsHash(txTo);
+    hashJoinSplits = GetJoinSplitsHash(txTo);
+}
+
+SigVersion SignatureHashVersion(const CTransaction& txTo)
+{
+    if (txTo.fOverwintered) {
```

Comment:
I'm a little nervous that it seems you can create a transaction with the `fOverwintered` flag being true, although the relevant bit in the header is not set. I wonder if it's better this be a function that checks this bit, 
at least in the SignatureHash function

Reply:
The header bit will always be set. `tx.nVersion` is not the header; the header is constructed from `tx.fOverwintered` and `tx.nVersion` (see `txTo.GetHeader()` further down).

Reply:
This also ties into the issues I've raised around consensus rules that assume `tx.nVersion` can be negative:

- https://github.com/zcash/zcash/pull/2925#discussion_r167231233
- https://github.com/zcash/zcash/pull/2925#discussion_r167231719

Reply:
^ Links don't open the specific comments any more, I can no longer see what this was referring to.

Reply:
Gah, that's a GitHub problem where the comment still exists, but because it is outdated it is rolled up by default, and can't be used as a link anchor.